Convert ReactPropConstantsTest to Kotlin

thiagobrez opened the pull request:
<!-- Thanks for submitting a pull request! We appreciate you spending the time to work on these changes. Please provide enough information so that others can review your pull request. The three fields below are mandatory. -->

## Summary:

<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->

Converts ReactPropConstantsTest to Kotlin, as requested in https://github.com/facebook/react-native/issues/38825

## Changelog:

<!-- Help reviewers and the release process by writing your own changelog entry.

Pick one each for the category and type tags:

[ANDROID|GENERAL|IOS|INTERNAL] [BREAKING|ADDED|CHANGED|DEPRECATED|REMOVED|FIXED|SECURITY] - Message

For more details, see:
https://reactnative.dev/contributing/changelogs-in-pull-requests
-->

[INTERNAL] [CHANGED] - Convert ReactPropConstantsTest to Kotlin
[INTERNAL] [CHANGED] - Add null-check to ViewManagersPropertyCache.java

## Test Plan:

<!-- Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots / videos if the pull request changes the user interface. -->

1. Run `./gradlew :packages:react-native:ReactAndroid:test`.
2. All tests should pass.

thiagobrez commented on packages/react-native/ReactAndroid/src/main/java/com/facebook/react/uimanager/ViewManagersPropertyCache.java:
```diff
@@ -393,6 +393,10 @@ public BoxedColorPropSetter(ReactProp prop, Method setter) {
    */
   /*package*/ static Map<String, PropSetter> getNativePropSettersForShadowNodeClass(
       Class<? extends ReactShadowNode> cls) {
+    if(cls == null) {
+      return new HashMap<>();
+    };
+
```
Added null-check because `cls` was coming `null` when calling this function recursively with `cls.getSuperClass()` in [lines 410-413](https://github.com/facebook/react-native/pull/39005/files#diff-26375553a8545f236d50426a9d97c4933eff4e4addd7070166b4527e4dda9cc3R410-R413)

cortinico replied:
Why is this change needed? Was one of your test failing?

thiagobrez replied:
@cortinico Exactly, `testNativePropsIncludeCorrectTypes` was failing with a NullPointerException when instantiating the UIManagerModule:

```kotlin
val uiManagerModule = UIManagerModule(reactContext, viewManagers, 0)
```

I tracked it all the way down to ViewManagersPropertyCache.java, in `getNativePropSettersForShadowNodeClass`:

When this function is called recursively (in lines 410-413) with `cls.getSuperclass()`, the superclass might be null, causing `cls.getInterfaces()` (line 400) to crash.

```java
for (Class iface : cls.getInterfaces()) {
  ...
}
```

Let me know if you believe there is a better way to solve this!

cortinico commented on packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:
```diff
@@ -0,0 +1,153 @@
+    val viewManagers = listOf<ViewManager<*, *>>(ViewManagerUnderTest())
+    val reactContext = ReactApplicationContext(RuntimeEnvironment.getApplication())
+    val uiManagerModule = UIManagerModule(reactContext, viewManagers, 0)
+    val constants: Map<*, *>? = valueAtPath(uiManagerModule.constants, "SomeView", "NativeProps")
```
```suggestion
    val constants: Map<*, *> = valueAtPath(uiManagerModule.constants, "SomeView", "NativeProps")
```

cortinico commented on packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:
```diff
@@ -0,0 +1,153 @@
+  }
+
+  companion object {
+    private fun valueAtPath(nestedMap: Map<String, *>?, vararg keyPath: String): Map<*, *>? {
```
```suggestion
    private fun valueAtPath(nestedMap: Map<String, *>, vararg keyPath: String): Map<*, *> {
```

cortinico reviewed (changes requested):
Hi @thiagobrez
Thanks for sending this over!

Can I ask you to fix the warnings that are causing the CI to fail:

```
> Task :packages:react-native:ReactAndroid:compileDebugUnitTestKotlin FAILED
e: warnings found and -Werror specified
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:50:21 Parameter 'v' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:50:31 Parameter 'value' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:54:20 Parameter 'v' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:54:30 Parameter 'value' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:58:22 Parameter 'v' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:58:32 Parameter 'value' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:62:23 Parameter 'v' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:62:33 Parameter 'value' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:66:23 Parameter 'v' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:66:33 Parameter 'value' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:70:26 Parameter 'v' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:70:36 Parameter 'value' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:74:25 Parameter 'v' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:74:35 Parameter 'value' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:78:22 Parameter 'v' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:78:32 Parameter 'value' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:82:20 Parameter 'v' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:82:30 Parameter 'value' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:86:27 Parameter 'v' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:86:37 Parameter 'index' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:86:49 Parameter 'value' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:90:25 Parameter 'v' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:90:35 Parameter 'index' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:90:47 Parameter 'value' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:94:30 Parameter 'v' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:94:40 Parameter 'index' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:94:52 Parameter 'value' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:98:23 Parameter 'v' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:98:33 Parameter 'value' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:105:28 Parameter 'v' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:105:38 Parameter 'index' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:105:50 Parameter 'value' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:148:34 Unchecked cast: Any? to Map<String, *>

> Task :packages:react-native:ReactAndroid:compileReleaseUnitTestKotlin FAILED
e: warnings found and -Werror specified
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:50:21 Parameter 'v' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:50:31 Parameter 'value' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:54:20 Parameter 'v' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:54:30 Parameter 'value' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:58:22 Parameter 'v' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:58:32 Parameter 'value' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:62:23 Parameter 'v' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:62:33 Parameter 'value' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:66:23 Parameter 'v' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:66:33 Parameter 'value' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:70:26 Parameter 'v' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:70:36 Parameter 'value' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:74:25 Parameter 'v' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:74:35 Parameter 'value' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:78:22 Parameter 'v' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:78:32 Parameter 'value' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:82:20 Parameter 'v' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:82:30 Parameter 'value' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:86:27 Parameter 'v' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:86:37 Parameter 'index' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:86:49 Parameter 'value' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:90:25 Parameter 'v' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:90:35 Parameter 'index' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:90:47 Parameter 'value' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:94:30 Parameter 'v' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:94:40 Parameter 'index' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:94:52 Parameter 'value' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:98:23 Parameter 'v' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:98:33 Parameter 'value' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:105:28 Parameter 'v' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:105:38 Parameter 'index' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:105:50 Parameter 'value' is never used
w: file:///root/react-native/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/uimanager/ReactPropConstantsTest.kt:148:34 Unchecked cast: Any? to Map<String, *>
```

as those are tests you'll have to `@Suppress` some of the warnings at the class level probably

thiagobrez commented:
@cortinico Of course, I wasn't sure if suppressing was recommended - sending fixes soon!

thiagobrez commented:
@cortinico Hey, just sent the following fixes:

- Suppressed the unused variables at the inner class level (`ViewManagerUnderTest`). Since we're just implementing a dummy class for the test, it seemed like a better idea than suppressing every individual function.

- Implemented the suggested changes, plus rewrote a few things in `valueAtPath` to fix the unchecked cast in I had in `value?.get(key) as? Map<String, *>`. Also, replaced `assertThat` with `require` since these were runtime validations. Because of that, I could remove the `error()` call and get a cleaner return type for the function.

Let me know what you think :)

cortinico commented on packages/react-native/ReactAndroid/src/main/java/com/facebook/react/uimanager/ViewManagersPropertyCache.java:
```diff
@@ -393,6 +393,10 @@ public BoxedColorPropSetter(ReactProp prop, Method setter) {
   /*package*/ static Map<String, PropSetter> getNativePropSettersForShadowNodeClass(
       Class<? extends ReactShadowNode> cls) {
+    if(cls == null) {
+      return new HashMap<>();
```
```suggestion
      return EMPTY_PROPS_MAP;
```

cortinico reviewed (approved):
> Let me know what you think :)

Changes looks great to me 👍 We can import and merge this one now
